File: 503-next-greater-element-ii/503-next-greater-element-ii.py
class Solution:
    def nextGreaterElements(self, A):
        stack, res = [], [-1] * len(A)
        for i in list(range(len(A)))*2:
            while stack and (A[stack[-1]] < A[i]):
                res[stack.pop()] = A[i]
            stack.append(i)
        return res
            

File: 503-next-greater-element-ii/503-next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int[] arr = new int[nums.length * 2];
        int n = nums.length;
        for(int i = 0; i < arr.length; i++) arr[i] = nums[i % n];
        
        int[] res = new int[nums.length];
        Arrays.fill(res, -1);
        Stack<Integer> stack = new Stack<>();
        for(int i = 0; i < arr.length; i ++) {
            while(!stack.isEmpty() && arr[stack.peek()] < arr[i]) {
                res[stack.pop() % n] = arr[i];
            }
            stack.push(i);
        }
        return res;
    }
}